admin/c2cgeoportal_admin/views/themes_ordering.py

```diff
@@ -0,0 +1,104 @@
+from pyramid.view import view_config
+from pyramid.httpexceptions import HTTPFound
+
+import colander
+from c2cgeoform.schema import GeoFormSchemaNode
+from c2cgeoform.views.abstract_views import AbstractViews
+from deform import ValidationFailure
+from sqlalchemy.sql.expression import literal_column
+
+from c2cgeoportal_commons.models.main import Theme
+
+from c2cgeoportal_admin.widgets import ThemeOrderWidget, ChildrenWidget
+
+
+class ThemeOrderSchema(GeoFormSchemaNode):
+
+    def objectify(self, dict_, context=None):
+        context = self.dbsession.query(Theme).get(dict_['id'])
+        context = super().objectify(dict_, context)
+        return context
+
+
+@colander.deferred
+def treeitems(node, kw):
+    return kw['dbsession'].query(Theme, literal_column('0')). \
+        order_by(Theme.ordering, Theme.name)
+
+
+def themes_validator(node, cstruct):
+    for dict_ in cstruct:
+        if not dict_['id'] in [item.id for item, group in node.treeitems]:
+            raise colander.Invalid(
+                node,
+                'Value {} does not exist in table {} or is not allowed to avoid cycles'.
+                format(dict_['treeitem_id'], Theme.__tablename__))
+
+
+class ThemesOrderingSchema(colander.MappingSchema):
+    themes = colander.SequenceSchema(
+        ThemeOrderSchema(
+            Theme,
+            includes=['id', 'ordering'],
+            name='theme',
+            widget=ThemeOrderWidget()
+        ),
+        name='themes',
+        treeitems=treeitems,
+        validator=themes_validator,
+        widget=ChildrenWidget(add_subitem=False, category='structural')
+    )
+
+
+class ThemesOrdering(AbstractViews):
+
+    _base_schema = ThemesOrderingSchema()
+
+    @view_config(route_name='layertree_ordering',
+                 request_method='GET',
+                 renderer='../templates/edit.jinja2')
+    def edit(self):
+        form = self._form()
+        dict_ = {
+            'themes': []
+        }
+        for theme, group in form.schema['themes'].treeitems:
+            dict_['themes'].append(
+                form.schema['themes'].children[0].dictify(theme)
+            )
+        rendered = form.render(dict_,
+                               request=self._request,
+                               actions=[])
+        return {
+            'form': rendered,
+            'deform_dependencies': form.get_widget_resources()
+        }
+
+    @view_config(route_name='layertree_ordering',
+                 request_method='POST',
+                 renderer='../templates/edit.jinja2')
+    def save(self):
+        try:
+            form = self._form()
+            form_data = self._request.POST.items()
+            self._appstruct = form.validate(form_data)
+            with self._request.dbsession.no_autoflush:
+                for dict_ in self._appstruct['themes']:
+                    obj = form.schema['themes'].children[0].objectify(dict_)
+            self._obj = self._request.dbsession.merge(obj)
```

Comment:
I don't understand why we have to objectify many themes but just to merge one, the last.

Reply:
good catch, self._obj make non sense here, we should merge all themes independently.